File: Vandame/matlab_code/E2_repetition_sound.m
compartement = struct(...
    'Connectivity', 8, ...
    'ImageSize', Data.spectroinfo.s, ...
    'NumObjects', [], ...
    'PixelIdxList', [] );
RIF = 100;
fft_mean = cell(max(clust),1);
time  =cell(max(clust),1);
for i = 1: max(clust)
    fft_mean{i}=zeros(Data.spectroinfo.s(2)*10,1);
end

for folder = 1 : NB_folder
    for file = 1 : NB_file
        beginning =  min(find(all_index(:,1) == folder & all_index(:,2) == file));
        endd =  max(find(all_index(:,1) == folder & all_index(:,2) == file));
        
        
        T =  beginning:endd;
        for i = 1 : max(clust)
            obj = find(clust == i);
            obj = obj(logical(sum(obj == T,2)));
            if ~isempty(obj)
                compartement.PixelIdxList = all_obj(obj);
                compartement.NumObjects = length(obj(:));
                center = regionprops(compartement,'centroid');
                center = cat(1,center.Centroid);
                center = center(:,1);
                time{i} = zeros(Data.spectroinfo.s(2),1);
                time{i}(round(center)) = 1;
                fft_mean{i}=fft_mean{i}+conv(abs(fft(time{i},Data.spectroinfo.s(2)*10)),ones(RIF,1),'same');
            end
        end
        
    end
end
peak = cell(max(clust),1);
location = cell(max(clust),1);
Frequel = 0:1/(Data.spectroinfo.s(2)*10):1/2.';
Frequel = Frequel(1:floor(end/10));
 warning ('off','signal:findpeaks:largeMinPeakHeight');

for i = 1:max(clust)
    fft_mean{i}=fft_mean{i}/norm(fft_mean{i});
    fft_mean{i}=fft_mean{i}(1:floor(end/2));
    fft_mean{i} = fft_mean{i}(1:floor(end/10));
    
    [peak{i},location{i}]=findpeaks(fft_mean{i},1,...
        'MinPeakHeight',7*10^-3,...
        'MinPeakDistance',1000);

    peak{i}(Frequel(location{i})<4/Data.spectroinfo.s(2)) = [];
    location{i}(Frequel(location{i})<4/Data.spectroinfo.s(2)) = [];


end
%%
clear center compartement beginning endd f fft_mean i j k peak RIF t tab1 time tabgp T obj folder file

File: Vandame/ChatGPT_MATtoPY/E2_repetition_sound.py
import numpy as np
from scipy.signal import find_peaks

compartement = {'Connectivity': 8, 'ImageSize': Data.spectroinfo.s, 'NumObjects': None, 'PixelIdxList': None}
RIF = 100
fft_mean = [np.zeros(Data.spectroinfo.s[1] * 10) for _ in range(max(clust))]
time = [np.zeros(Data.spectroinfo.s[1]) for _ in range(max(clust))]
peak = [None] * max(clust)
location = [None] * max(clust)

for folder in range(NB_folder):
    for file in range(NB_file):
        T = np.arange(min(np.where((all_index[:, 0] == folder) & (all_index[:, 1] == file))[0]),
                      max(np.where((all_index[:, 0] == folder) & (all_index[:, 1] == file))[0]) + 1)

        for i in range(max(clust)):
            obj = np.where(clust == i + 1)[0]
            obj = obj[np.any(obj[:, None] == T, axis=1)]
            if obj.size > 0:
                compartement['PixelIdxList'] = all_obj[obj]
                compartement['NumObjects'] = len(obj)
                center = regionprops(compartement, 'centroid')[0]['Centroid'][0]
                time[i] = np.zeros(Data.spectroinfo.s[1])
                time[i][round(center)] = 1
                fft_mean[i] += np.convolve(np.abs(np.fft.fft(time[i], Data.spectroinfo.s[1] * 10)), np.ones(RIF),
                                           mode='same')

        fft_mean_norm = [fft / np.linalg.norm(fft[:fft.size // 2][:fft.size // 10]) for fft in fft_mean]

        for i, fft in enumerate(fft_mean_norm):
            peaks, locations = find_peaks(fft, height=7 * 10 ** -3, distance=1000)
            freq_locations = Frequel[locations]
            mask = freq_locations >= 4 / Data.spectroinfo.s[1]
            peak[i] = peaks[mask]
            location[i] = locations[mask]

# Clear variables
compartement = None
RIF = None